compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -561,31 +560,69 @@ static bool isTransposeOp(linalg::LinalgOp linalgOp) {
   if (linalgOp.hasDynamicShape()) {
     return false;
   }
-
-  // Check that at least one input operands is transposed.
-  bool hasPermutation = false;
-  for (auto indexMap : linalgOp.getIndexingMapsArray()) {
-    if (isSharedMemTranspose(indexMap)) {
-      hasPermutation = true;
-    }
-  }
-  return hasPermutation;
+  return true;
 }
 
 static LogicalResult setTransposeConfig(func::FuncOp entryPoint,
+                                        linalg::GenericOp linalgOp,
                                         Operation *op) {
+  if (!checkTransposePreconditions(linalgOp)) {
+    return failure();
+  }
+
+  // To simplify logic, we only consider linalg ops with transposes who's
+  // outputs maps are identities
+  for (int i = linalgOp.getNumInputs();
+       i < linalgOp.getIndexingMapsArray().size(); ++i) {
+    auto indexMap = linalgOp.getIndexingMapsArray()[i];
+    if (!indexMap.isIdentity()) {
+      return failure();
+    }
+  }
+
+  // Determine which operands are transposed.
+  SmallVector<int64_t, 4> transposedOperandIndices;
+  for (auto indexMapPair : llvm::enumerate(linalgOp.getIndexingMapsArray())) {
+    if (isSharedMemTranspose(indexMapPair.value())) {
+      transposedOperandIndices.push_back(indexMapPair.index());
+    }
+  }
+
+  if (transposedOperandIndices.empty()) {
+    return failure();  // No shared mem transposes.
+  }
+
+  // Determine the fastest moving dimensions for the source/destination indices
+  // of each transpose. These inform the tile sizes.
+  int64_t outputFastestDim = linalgOp.getNumLoops() - 1;
+  int64_t inputFastestDim =
+      linalgOp.getIndexingMapsArray()[transposedOperandIndices[0]]
+          .getDimPosition(outputFastestDim);
+  // Ensure the other transposed operands match
+  for (int i = 1; i < transposedOperandIndices.size(); ++i) {
+    if (inputFastestDim !=
+        linalgOp.getIndexingMapsArray()[transposedOperandIndices[0]]
```

Comment:
looks like you are always checking the first input. Can you add a negative test for this case?